### PR TITLE
Rework credential management on LwM2M client library 

### PR DESCRIPTION
**Pull request by SeppoTakalo**
Store credentials and settings

LwM2M client utilities library will now independently handle
storage of credentials to modem and other security settings
to permanent storage. This includes various resources from
LwM2M objects server (ID 1) and security (ID 0).

Client utilities installs load_credentials hook into
client contexts and handles modem keys on this hook.
There is a side effect, that we might need to turn the modem
offline, if keys need to be updated.

For permanent storage, we use Zephyr's settings API.
We install create&destroy hooks into LwM2M engine for
security and server objects. When such object is created,
we install post-write hook on each of the resources we care about.

From Server object, we monitor resources:
* ID 0, Short Server ID
* ID 1, Lifetime

From Security object, we monitor:
* ID 0, LWM2M Server URI
* ID 2, Security mode
* ID 3, PSK Identity
* ID 4, Server Public key
* ID 10, Short Server ID

When we receive this data from Bootstrap server, our hooks
ensure that creted objects end up written in the permanent storage.
When client connection is started, the PSK from security object is
written to the modem. PSK is never written into settings storage
and after a reboot, it is not readable anymore.

On a reboot, we load these stored objects from settings storage,
and create them into LwM2M engine. Separate API has published
to allow application to check, whether we have client certificates
already, or if we need to bootstrap.

This PR contains changes from:
https://github.com/nrfconnect/sdk-nrf/pull/7023

### CODE REVIEW

**Review by jheiskan81 — APPROVED** [review]

**Comment by VeijoPesonen** [comment]
I propose that we would introduce our own overlay files for Leshan and Coiote configurations. Both could have bootstrap and queue mode configs but as commented out. Just add instructions there that by uncommenting those sections both features can be turned on.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c [hunk 1]
```diff
@@ -7,20 +7,330 @@
...
+
+	snprintk(pathstr, sizeof(pathstr), "0/%d/%d", ctx->sec_obj_inst, res_id);
+	ret = lwm2m_engine_get_res_data(pathstr, &cred, &cred_len, &cred_flags);
+	if (ret < 0) {
+		LOG_ERR("Unable to get resource data for '%s'", log_strdup(pathstr));
+		return ret;
+	}
+
+	/* Convert binary PSK key to hex format, which is expect by modem */
+if (type == MODEM_KEY_MGMT_CRED_TYPE_PSK) {
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c:55]: Indentation.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c [hunk 2]
```diff
@@ -7,20 +7,330 @@
...
+	/* Never overwrite bootstrap keys. */
+	if (exist &&
+	    (have_permanently_stored_keys
+	    || ctx->tls_tag == CONFIG_LWM2M_CLIENT_UTILS_BOOTSTRAP_TLS_TAG)) {
+		LOG_DBG("Existing credentials found on modem");
+		return 0;
+	}
+
+	if (lte_lc_func_mode_get(&fmode)) {
+		LOG_ERR("Failed to read modem functinal mode");
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c:102]: Typo

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]

**Review by rlubos — COMMENTED** [review]
I think this change merits an entry in the changelog.

**Comment by rlubos** [comment]
@umapraseeda Please reassign if lwm2m docs should be handled by someone else.

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 3]
```diff
@@ -572,9 +359,15 @@ void main(void)
 #endif
 
 	while (true) {
-		lwm2m_rd_client_start(&client, endpoint_name, flags, rd_client_event, NULL);
+		if (IS_ENABLED(CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP) &&
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply samples/nrf9160/lwm2m_client/src/main.c:362]: Actually, this is already checked there, so I can drop the `IS_ENABLED()` test.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c (hunk 2 above)
Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c]: fixed.

File: samples/nrf9160/lwm2m_client/overlay-ci.conf [hunk 4]
```diff
@@ -1,2 +1,2 @@
 # LwM2M server needs to be set to pass CI
-CONFIG_APP_LWM2M_SERVER="passci"
+CONFIG_LWM2M_CLIENT_UTILS_SERVER="passci"
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply samples/nrf9160/lwm2m_client/overlay-ci.conf]: Removed the overlay-ci.conf

File: samples/nrf9160/lwm2m_client/src/main.c [hunk 5]
```diff
@@ -84,6 +83,16 @@ static int lwm2m_setup(void)
...
 	lwm2m_init_security(&client, endpoint_name);
 
+	if (IS_ENABLED(CONFIG_LWM2M_DTLS_SUPPORT)) {
+		/* Write hard-coded PSK key to engine */
+		char buf[1 + sizeof(CONFIG_APP_LWM2M_PSK) / 2];
+		size_t len = hex2bin(CONFIG_APP_LWM2M_PSK, sizeof(CONFIG_APP_LWM2M_PSK) - 1, buf,
+				     sizeof(buf));
+
+		/* First security instance is the right one, because in bootstrap mode, */
+		/* it is the bootstrap PSK. In normal mode, it is the server key */
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply samples/nrf9160/lwm2m_client/src/main.c:93]: fixed.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c [hunk 6]
```diff
@@ -8,19 +8,329 @@
...
+		LOG_INF("Need to write credentials, switch LTE offline");
+		lte_lc_offline();
+	}
+
+	ret = write_credential_type(ctx, SECURITY_CLIENT_PK_ID, MODEM_KEY_MGMT_CRED_TYPE_IDENTITY);
+	if (ret) {
+		goto out;
+	}
+
+	ret = write_credential_type(ctx, SECURITY_SECRET_KEY_ID, MODEM_KEY_MGMT_CRED_TYPE_PSK);
```
[the remark this answers is not in the document]
Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c:143]: `ret` was supposed to be returned. but I just realized that I have added functionality after this one, so it did not match the intention any more.
Fixed now

**Review by VeijoPesonen — APPROVED** [review]
Looks good to me

**Review by umapraseeda — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by rlubos** [comment]
@SeppoTakalo sorry for the delay, I've lost the PR from my radar, could you please rebase to resolve the conflict?

**Pull request merged**